### ISSUE TITLE
s3.getObject().createReadStream().pipe() hangs in Node v0.10.25

### ISSUE THREAD

**Issue by dzielin** [aws/aws-sdk-js#928]
The symptoms are the same as listed here for a previous issue (https://github.com/aws/aws-sdk-js/issues/163).

It seems that intermittently, S3's getObject() hangs and fails to return a response.  When this happens, it prevents all other calls to getObject() from being returned as well, while the rest of the node application continues functioning without reporting any errors.

Supposedly the TLS regression was fixed in v0.10.20, so I shouldn't be experiencing this error.  I do not see this happen terribly often (once every few weeks), so it has been difficult to diagnose.  Even when I fire off many API requests, it is impossible to manually recreate this issue.

My usage of this API is quite simple.  Something to the effect of:
```
var params = {Bucket: 'XXXXXXXXXXXXX', Key: 'XXXXXXXXXXXX'};
res.set('Content-Type', 'application/octet-stream');
s3.getObject(params).createReadStream().pipe(res);
```

Like I said, this is an intermittent problem.  If I restart the node application and try the same exact request, the file will download without issue.

If this is not in fact an issue with the S3 API, I would be quite grateful if someone could point me in the right direction for diagnosing the potential Node.js TLS re-regression.

**Comment by chrisradek** [comment]
@dzielin 
Can you tell me what version of the SDK you're using? What region is your bucket in, and have you noticed any patterns, like seeing this issue with larger files?

It's possible this is related to https://github.com/aws/aws-sdk-js/issues/827 as well, which was worked-around by moving to node 0.12.x+. If you can provide a fuller example of your test case, I can try looking into it more to determine if it is an SDK issue or a node issue.

**Comment by dzielin** [comment]
I believe this was on v2.2.41 of the SDK.  I've since moved to v2.2.43.  I have also updated Node to the latest build (5.8.0).  So far, I have yet to see this issue reappear.  But I can't say this with any amount of certainty for several more weeks.

My bucket is in the us-east-1 region.  The files I've been working with are typically 1MB to 20MB... but there are exceptions on both ends of the spectrum.  The most recent failure was a 3MB PDF.  It has been known to happen with small files (as small as 3 bytes; a test file).  Restarting the node application always fixes the issue and I can immediately execute the exact same request with no troubles.  I have not been able to identify any patterns as to when this happens (which makes attempting to debug it all the more difficult) aside from the fact that it's infrequent.

The issue presents itself after some time, typically at least a couple days of running the node application.  Heap size doesn't seem to be growing unexpectedly during this time... in fact it stabilizes quite nicely and remains within +/- 5MB during runtime.  Nothing times out, it will just hang for hours or even days until the application is restarted.

For my application, I am doing a database lookup in my Express route, so the actual code looks more like this:
```
var express = require('express');
var connection = require('../mysqlconnection');
var router = express.Router();
var AWS = require('aws-sdk');

var s3 = new AWS.S3();

router.get('/:id([0-9]{1,10})', function (req, res, next) {
    var id = req.params.id;

	connection.execute('SELECT file_name, file_type, s3_id FROM file_map WHERE id=?', [id], function (err, rows) {
		if (!err && !isEmptyObject(rows)) {
			var file_type = rows[0].file_type;
            var file_name = rows[0].file_name;
			var s3_id = rows[0].s3_id;
			
            var params = {Bucket: req.app.get('awsbucket'), Key: s3_id}; // The value for 'awsbucket' is pulled via app.js from config.js
			
            res.set('Content-Type', 'application/octet-stream');
            res.set('Content-Disposition', 'attachment; filename=\"' + file_name + '\"');
			
            s3.getObject(params).createReadStream().pipe(res);
        } else {
			res.json({success: false, message: "No matching files found."});
		}
    });
});
```

I have been able to verify that the database connection is still solid after S3 hangs by running other routes that utilize the same connection object without issue.  And, because I've been able to restart the node application and successfully execute the S3 getObject, I can say with certainty that it isn't an issue of missing or inaccurate data in the relational database.

**Comment by chrisradek** [comment]
@dzielin 
Thanks for the sample, I'll try to reproduce on my end, though it sounds like that may take some time.

It's interesting that all subsequent getObject calls hang as well. Nodejs 0.10.x by default sets maxSockets to 5, and I'm wondering if there is an issue where a request hangs and that socket never gets freed.

Please keep us posted with any issues related to this you find after upgrading your version of node!

**Comment by dzielin** [comment]
Do you know of an easy way to display sockets in use by node for debug purposes? I'd be interested in taking a look at that on my end, but it would be preferable to do it within the node app and just log it to console occasionally.  In the meantime I'll see if I can do it at the OS level.

**Comment by kichooo** [comment]
I believe that I encounter the same problem. On production server, we create lots of medium (20-50MB) files on S3. After a while, some nodejs processes will stop putting new files, and even hang the express app. In the same time, other processes in the same deployment work nice.

**Comment by dzielin** [comment]
@kichooo That sounds exactly like our issue.  I believe @chrisradek is correct in noting the default maxSockets.  We fixed it by upgrading to Node v5.8.0.  This may be because v5.8.0 doesn't have a finite default maxSockets, or it may be because it is properly freeing connections.

While I haven't been able to diagnose the source of the fix, I would suggest setting your maxSockets manually.  Something like `https.globalAgent.maxSockets = Infinity;`

**Comment by chrisradek** [comment]
@kichooo 
You can set the maxSockets per client by doing the following:
```javascript
// Example using s3
var https = require('https');
// ...
var s3 = new AWS.S3({
  httpOptions: {
    agent: new https({
      maxSockets: 1000 // Number will depend on your server's specs
      rejectUnauthorized: true
    })
  }
});
```
Closing the issue since upgrading Node seemed to fix it. Feel free to comment if defining a different value for `maxSockets` doesn't fix the problem.
